src/Features/Core/Portable/Completion/Providers/ImportCompletionProvider/ExtensionMethodImportCompletionHelper.cs

```diff
@@ -265,16 +263,91 @@ static ImmutableArray<string> AttachComplexTypes(ImmutableArray<string> receiver
             }
         }
 
-        private static void GetExtensionMethodItemsForSymbolsFromDifferentCompilation(
+        private static async Task<ImmutableArray<IMethodSymbol>?> GetExtensionMethodSymbolsFromProjectAsync(
+            Project project,
+            bool isCurrentProject,
+            bool forceIndexCreation,
+            SemanticModel semanticModel,
+            IImportCompletionCacheService<CacheEntry, object> cacheService,
+            ITypeSymbol receiverTypeSymbol,
+            ImmutableArray<string> receiverTypeNames,
+            int position,
+            ISet<string> namespaceFilter,
+            ConcurrentDictionary<ITypeSymbol, bool> checkedReceiverTypes,
+            CancellationToken cancellationToken)
+        {
+            var cacheEntry = await GetCacheEntryAsync(project, !forceIndexCreation, cacheService, cancellationToken).ConfigureAwait(false);
+            if (!cacheEntry.HasValue)
+            {
+                // Returns null to indicate index not ready
+                return null;
+            }
+
+            if (!cacheEntry.Value.ContainsExtensionMethod)
+            {
+                return ImmutableArray<IMethodSymbol>.Empty;
+            }
+
+            var currentAssembly = semanticModel.Compilation.Assembly;
+            var filter = CreateAggregatedFilter(receiverTypeNames, cacheEntry.Value);
+            var compilation = await project.GetRequiredCompilationAsync(cancellationToken).ConfigureAwait(false);
+            var assembly = compilation.Assembly;
+            var internalsVisible = currentAssembly.IsSameAssemblyOrHasFriendAccessTo(assembly);
+
+            var matchingMethodSymbols = GetPotentialMatchingSymbolsFromAssembly(
+                compilation.Assembly, filter, namespaceFilter, internalsVisible, cancellationToken);
+
+            return isCurrentProject
+                ? GetExtensionMethodsForSymbolsFromSameCompilation(
+                    position, semanticModel, receiverTypeSymbol, matchingMethodSymbols, checkedReceiverTypes, cancellationToken)
+                : GetExtensionMethodsForSymbolsFromDifferentCompilation(
+                    position, semanticModel, receiverTypeSymbol, matchingMethodSymbols, checkedReceiverTypes, cancellationToken);
```

Comment:
yup.  needing to pass huge swaths of parameters along in parallel is a real red flag that this should be a helper class.